File: src/converters/MastercardConverter.ts
import { FileConverter, TransactionRecord } from './FileConverter';

export class MastercardConverter implements FileConverter {
  private readonly MC_LINE_REGEX = /^MC\|(\d{2}-\d{2}-\d{4})\|(\d+\.\d{2})\|(.{30})\|(\d{16})\|(\w{10})$/;

  validate(fileContent: string): boolean {
    const lines = fileContent.split('\n');
    return lines.every(line => this.MC_LINE_REGEX.test(line.trim()));
  }

  parse(fileContent: string): TransactionRecord[] {
    if (!this.validate(fileContent)) {
      throw new Error('Invalid Mastercard file format');
    }

    const lines = fileContent.split('\n');
    return lines.map(line => {
      const [_, date, amount, merchant, card, txId] = line.trim().match(this.MC_LINE_REGEX) || [];
      
      return {
        transactionDate: this.parseDate(date),
        amount: parseFloat(amount),
        merchantName: merchant.trim(),
        cardNumber: card,
        transactionId: txId
      };
    });
  }

  private parseDate(dateStr: string): Date {
    // Assuming date format is DD-MM-YYYY
    const [day, month, year] = dateStr.split('-').map(num => parseInt(num));
    return new Date(year, month - 1, day);
  }
} 

File: converters/mastercard_converter.py
from datetime import datetime
import re

class MastercardConverter:
    def __init__(self):
        # Updated pattern to be more flexible with whitespace
        self.line_pattern = re.compile(r'^MC\|(\d{2}-\d{2}-\d{4})\|(\d+\.\d{2})\|(.+?)\|(\d{16})\|(\w{10})$')

    def parse(self, file_content: str) -> list:
        transactions = []
        for line in file_content.strip().split('\n'):
            match = self.line_pattern.match(line.strip())
            if not match:
                raise ValueError(f"Invalid Mastercard format: {line}")
            
            date_str, amount, merchant, card, tx_id = match.groups()
            
            # Parse date (DD-MM-YYYY format)
            date = datetime.strptime(date_str, '%d-%m-%Y')
            
            transactions.append({
                'transaction_date': date,
                'amount': float(amount),
                'merchant_name': merchant.strip(),
                'card_number': card,
                'transaction_id': tx_id
            })
        
        return transactions 